spring-cloud-app-broker-core/src/main/java/org/springframework/cloud/appbroker/workflow/instance/BackingServicesUpdateValidatorService.java

```diff
@@ -0,0 +1,79 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.appbroker.workflow.instance;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import reactor.core.publisher.Mono;
+
+import org.springframework.cloud.appbroker.deployer.BackingService;
+
+/**
+ * Validates updates to backing services, and returns an updated copy of the list with the previousPlan set. This help
+ * only triggering plan update when previousPlan is set.
+ */
+public class BackingServicesUpdateValidatorService {
+
+	// Accept and return Mono to make it easier for AppDeploymentUpdateServiceInstanceWorkflow to provide us with previous and candidate backing services
+	public Mono<List<BackingService>> validatePlanUpdate(Mono<List<BackingService>> previousBackingServices,
+		Mono<List<BackingService>> candidateBackingServicesForUpdate) {
+		try {
+			return Mono.just(validateAndMutatePlanUpdate(previousBackingServices.block(),
+				candidateBackingServicesForUpdate.block()));
+		}
+		catch (IllegalArgumentException e) {
+			return Mono.error(e);
+		}
+	}
+
+	public List<BackingService> validateAndMutatePlanUpdate(List<BackingService> previousBackingServicesList,
+		List<BackingService> candidateBackingServicesList) {
+		return candidateBackingServicesList.stream()
+			.map(candidateBackingService -> {
+				BackingService matchingPreviousBackingService = getValidatedBackingService(candidateBackingService,
+					previousBackingServicesList);
+				if (candidateBackingService.getPlan().equals(matchingPreviousBackingService.getPlan())) {
+					return candidateBackingService;
+				}
+				else {
+					//Don't mutate the caller's BackingService, return a copy instead.
+					return BackingService.builder()
+						.backingService(candidateBackingService)
+						.previousPlan(matchingPreviousBackingService.getPlan())
+						.build();
+				}
+
+			})
+			.collect(Collectors.toList());
+	}
+
+	private BackingService getValidatedBackingService(BackingService candidate,
+		List<BackingService> previousBackingServicesList) {
+		List<BackingService> matchingPreviousBackingServices = previousBackingServicesList.stream()
+			.filter(previousBackingService ->
+				previousBackingService.getName().equals(candidate.getName()) &&
+					previousBackingService.getServiceInstanceName().equals(candidate.getServiceInstanceName()))
+			.collect(Collectors.toList());
+		if (matchingPreviousBackingServices.size() != 1) { // NOPMD
```

Comment:
Any reason to disallow this completely?

Reply:
The reason to reject such update is that other types of upgrades (reported in #293 and illustrated into `BackingServicesUpdateValidatorServiceTest` ) would probably not be properly fullfilled by the CloudFoundryAppDeployer such as:
* deleting removed backing services
* changing type of backing service (e.g. from mysql to postgresql)
* changing service instance name (i.e. renaming a service instance)

May be the case of adding new backing services could be supported in the future with additional work.

To keep it simple as a first step, I was favoring a "fail-fast" approach.